### PR TITLE
feat: support generating docs for mulitple languages and splitting by namespace

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
This is how the result might look like in the context of cdktf: https://github.com/cdktf/cdktf-provider-random/blob/split-up-docs/API.md

I'm requesting feedback early here, I'll add more docs when you folks are happy with the overall concept.

### CODE REVIEW

**Comment by RomainMuller** [comment]
I am pretty favorable to smaller, more focused files generally speaking. That naturally applies to documentation files, too.

---

One thing that can be interesting (optionally - I reckon that might be substantial additional complexity, possibly unwarranted), is to actually "collapse" the navigation tree when there some module-like item only contains a few module-like items... For example (this is all made up, obviously):

```
API.md -> (foo)foo.md -> (bar)bar.md -> (baz)baz.md -> [types within foo.bar.baz]
       -> (foo)foo.md -> (bar2)bar2.md -> [types within foo.bar2]
```

Could be rendered as:

```
API.md -> (foo.bar.baz)baz.md
       -> (foo.bar2)bar2.md
```

This improves the navigation path (fewer clicks, better discoverability) when there is a small enough number of submodules in a given module.

**Comment by DanielMSchmidt** [comment]
I'm not sure I understand your comment 100%, let me try to rephrase it. You'd like to add an option so that instead of `Api.md -> Typescript -> <namespaceA>` and `Api.md -> Python -> <namespaceB>` you'd have a `Api.md` with a `Typescript` section, so that you can diectly click on `<namespaceA>` from within there?

**Comment by RomainMuller** [comment]
I was "ignoring" the language specific section, because I think it's okay to single out languages (folks will want to be in the one language's context anyway).

The idea is if a namespace only contains a few items, instead of linking to a list page for that namespace, have links to all the `namespace.Item`s instead, so users don't have to hop through a dedicated page just for a short list of things.

**Review by ryparker — CHANGES_REQUESTED** [review]

File: src/cli.ts [hunk 1]
```diff
@@ -3,29 +3,85 @@ import * as yargs from 'yargs';
...
+    .option('output', { alias: 'o', type: 'string', required: false, desc: 'Output filename, the file type is automatically added. (defaults to API.md if format is markdown, and API.json if format is JSON and one language is passed, API.language.md and API.language.json if multilpe are passed)' })
     .option('format', { alias: 'f', default: 'md', choices: ['md', 'json'], desc: 'Output format, markdown or json' })
-    .option('language', { alias: 'l', default: 'typescript', choices: Language.values().map(x => x.toString()), desc: 'Output language' })
+    .option('language', { array: true, alias: 'l', default: ['typescript'], choices: Language.values().map(x => x.toString()), desc: 'Output language' })
     .option('package', { alias: 'p', type: 'string', required: false, desc: 'The name@version of an NPM package to document', defaultDescription: 'The package in the current directory' })
     .option('submodule', { alias: 's', type: 'string', required: false, desc: 'Generate docs for a specific submodule (or "root")' })
+    .option('split-by-submodule', { type: 'boolean', required: false, desc: 'Generate a separate file for each submodule' })
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply src/cli.ts:57]: Yeah, I think that makes sense :) The alternative would be to make two distinct entry points, one for the single-file generation and one that creates a directory, this way we could split up the args into two bucket and don't have this state where some make limited sense on one use-case or the other :)

File: src/cli.ts [hunk 2]
```diff
@@ -3,29 +3,85 @@ import * as yargs from 'yargs';
...
 
+type GenerateOptions = {
+  readme?: boolean;
+  language: Language;
+  submodule?: string;
+  allSubmodules?: boolean;
+  splitBySubmodules?: boolean;
+}
+
+async function generateForLanguage(docs: Documentation, format: 'md' | 'json', options: GenerateOptions, output = 'API') {
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply src/cli.ts:14]: Sure thing!

File: src/cli.ts [hunk 3]
```diff
@@ -3,29 +3,90 @@ import * as yargs from 'yargs';
...
+        ...options,
+        submodule: submodule.name,
+        allSubmodules: false,
+        header: { title: `\`${submodule.name}\` Submodule`, id: submodule.fqn },
+      });
+
+      fs.writeFileSync(`${submodule.name}.${submoduleSuffix}`, content.render());
+    }
+
+    fs.writeFileSync(`${output ?? 'API'}.${fileSuffix}`, await (await docs.toIndexMarkdown(submoduleSuffix, options)).render());
```
Comment (RomainMuller) [inline src/cli.ts:44]: Feels a bit weird to use `Sync` FS APIs while you're using `await` there... Can you flip to using the promise versions (either from `fs/promises` or `import { promises as fs } from 'fs';`)?

**Review by RomainMuller — APPROVED** [review]

**Comment by mrgrain** [comment]
@DanielMSchmidt @RomainMuller I think this accidentally introduced a breaking change.
This [line](https://github.com/cdklabs/jsii-docgen/pull/903/files#diff-fa8d4e24d8399e8350f1c8bad05df53e8032ea995835bf911507015e2db61cddR86) changes the behaviour of `--output` from requiring an extension to not needing one, causing something like `jsii-docgen -o API.md` to now create a file named `API.md.md`.

At least this needs to be a breaking release. But saying that, it's a very annoying change for `projen` since it breaks the CLI contract for previous versions. 

See: https://github.com/projen/projen/issues/2458

**Comment by DanielMSchmidt** [comment]
Oh no, I'll provide a fix as soon as possible

**Pull request merged**